File: Back/src/main/kotlin/com/raminagrobis/centraleachat/app/security/jwt/JWTTokenUtil.java
package com.raminagrobis.centraleachat.app.security.jwt;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.core.env.Environment;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

@Component
public class JWTTokenUtil {

    private static final long JWT_TOKEN_VALIDITY = 5 * 60 * 60;

    private String secret;

    private final SecretKey KEY;

    public JWTTokenUtil(Environment env) {
        this.secret = env.getProperty("jwt.secret");
        this.KEY = Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
    }

    public String generateToken(UserDetails user) {
        Map<String, Object> claims = new HashMap<>();

        return doGenerateToken(claims, user.getUsername());
    }

    public String getUsernameFromToken(String token) {
        return getClaimFromToken(token, Claims::getSubject);
    }

    public String getExpiration(String token) {
        return getClaimFromToken(token, Claims::getExpiration).toString();
    }

    public String getIssuedAt(String token) {
        return getClaimFromToken(token, Claims::getIssuedAt).toString();
    }

    public Boolean validateToken(String token, UserDetails user) {
        return getUsernameFromToken(token).equals(user.getUsername()) && !isTokenExpired(token);
    }


    private <T> T getClaimFromToken(String token, Function<Claims, T> claimsResolver) {
        Claims claims = getAllClaimsFromToken(token);

        return claimsResolver.apply(claims);
    }


    private String doGenerateToken(Map<String, Object> claims, String subject) {
        return Jwts.builder().setClaims(claims)
                .setSubject(subject)
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + JWT_TOKEN_VALIDITY * 1000))
                .signWith(this.KEY)
                .compact();
    }

    private Claims getAllClaimsFromToken(String token) {
        return Jwts.parserBuilder().setSigningKey(this.KEY).build().parseClaimsJws(token).getBody();
    }

    private Boolean isTokenExpired(String token) {
        return getClaimFromToken(token, Claims::getExpiration).before(new Date());
    }
}

File: Back/src/main/kotlin/com/raminagrobis/centraleachat/app/security/jwt/JWTTokenUtil.kt
package com.raminagrobis.centraleachat.app.security.jwt

import com.raminagrobis.centraleachat.domain.administration.model.Role
import com.raminagrobis.centraleachat.domain.connexion.adapter.IJWTTokenUtil
import com.raminagrobis.centraleachat.domain.connexion.model.Utilisateur
import com.raminagrobis.centraleachat.infra.utilisateur.UtilisateurRepo
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.core.env.Environment
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.function.Function
import javax.crypto.SecretKey

@Service
class JWTTokenUtil(env: Environment, val utilisateurRepo: UtilisateurRepo): IJWTTokenUtil {

    val JWT_TOKEN_VALIDITY = (5 * 60 * 60).toLong()

    var secret: String? = null

    var KEY: SecretKey? = null

    init {
        this.secret = env.getProperty("jwt.secret")
        this.KEY = Keys.hmacShaKeyFor(secret!!.toByteArray(StandardCharsets.UTF_8))
    }

    override fun generateToken(user: UserDetails): String {
        val claims: Map<String, Any> = HashMap()
        return doGenerateToken(claims, user.username)
    }

    fun getUsernameFromToken(token: String): String {
        return getClaimFromToken<String>(token, Function<Claims, String> { obj: Claims -> obj.subject })
    }

    fun getExpiration(token: String): String {
        return getClaimFromToken<Date>(token, Function<Claims, Date> { obj: Claims -> obj.expiration }).toString()
    }

    fun getIssuedAt(token: String): String {
        return getClaimFromToken<Date>(token, Function<Claims, Date> { obj: Claims -> obj.issuedAt }).toString()
    }

    fun validateToken(token: String): Boolean {
        return getUsernameFromToken(token) != null && !isTokenExpired(token)
    }


    fun <T> getClaimFromToken(token: String, claimsResolver: Function<Claims, T>): T {
        val claims: Claims = getAllClaimsFromToken(token)
        return claimsResolver.apply(claims)
    }


    fun doGenerateToken(claims: Map<String, Any?>, subject: String): String {
        return Jwts.builder().setClaims(claims)
            .setSubject(subject)
            .setIssuedAt(Date(System.currentTimeMillis()))
            .setExpiration(Date(System.currentTimeMillis() + JWT_TOKEN_VALIDITY * 1000))
            .signWith(this.KEY)
            .compact()
    }

    fun getAllClaimsFromToken(token: String): Claims {
        return Jwts.parserBuilder().setSigningKey(this.KEY).build().parseClaimsJws(token).getBody()
    }

    fun isTokenExpired(token: String): Boolean {
        return getClaimFromToken(token) { obj: Claims -> obj.expiration }.before(Date())
    }

    fun getUtilisateurFromToken(token : String) : Utilisateur {
        var utilisateur : Utilisateur
        val email = getUsernameFromToken(token)
        try {
            utilisateur = utilisateurRepo.findAdminByEmail(email)!!
        } catch (e : Exception){
            utilisateur = utilisateurRepo.findSocieteByEmail(email)!!
        }

        return utilisateur
    }

}
